core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,593 @@
+                .lookup_contact_info(peer, |ci| ci.alpenglow())
+                .flatten()
+            else {
+                continue;
+            };
+            state.peers.insert(
+                *peer,
+                PeerData {
+                    stake: stake,
+                    address: ag_addr,
+                    relative_toa: [None; 4],
+                },
+            );
+            state.total_staked += stake;
+        }
+        trace!(
+            "Prepared for slot {slot}, total stake is {}",
+            state.total_staked
+        );
+    }
+
+    /// Collects votes and changes states
+    fn listener_thread(
+        self_state: Arc<Mutex<SharedState>>,
+        cluster_info: Arc<ClusterInfo>,
+        mut epoch_specs: EpochSpecs,
+        socket: Arc<UdpSocket>,
+        command_sender: Sender<Command>,
+    ) {
+        socket
+            .set_read_timeout(Some(Duration::from_secs(1)))
+            .unwrap();
+        let id = cluster_info.id();
+        // Set aside enough space to fetch multiple packets from the kernel per syscall
+        let mut packets: Vec<Packet> = vec![Packet::default(); 1024];
+        loop {
+            // must wipe all Meta records to reuse the buffer
+            for p in packets.iter_mut() {
+                *p.meta_mut() = Meta::default();
+            }
+            // recv_mmsg should timeout in 1 second
+            let n = match recv_mmsg(&socket, &mut packets) {
+                Ok(n) => n,
+                Err(e) => {
+                    match e.kind() {
+                        std::io::ErrorKind::TimedOut | std::io::ErrorKind::WouldBlock => {
+                            0 // no packets received
+                        }
+                        _ => {
+                            error!("Got error {:?} in mock alpenglow RX socket operation, exiting thread", e.raw_os_error());
+                            return;
+                        }
+                    }
+                }
+            };
+            let mut state = self_state.lock().unwrap();
+            if state.should_exit {
+                return;
+            }
+            // we may have received no packets, in this case we can safely skip the rest
+            if n == 0 {
+                continue;
+            }
+            let elapsed = state.current_slot_start.elapsed();
+
+            let stake_60_percent = (state.total_staked as f64 * 0.6) as Stake;
+            let stake_80_percent = (state.total_staked as f64 * 0.8) as Stake;
+            for pkt in packets.iter().take(n) {
+                if pkt.meta().size < MOCK_VOTE_HEADER_SIZE {
+                    continue;
+                }
+                let Some(pkt_buf) = pkt.data(..) else {
+                    continue;
+                };
+                let vote_pkt = MockVotePacketHeader::from_bytes(pkt_buf);
+                let pk = Pubkey::new_from_array(vote_pkt.sender);
+                if vote_pkt.slot_number != state.current_slot {
```

Comment:
I wonder whether we need to keep the stats for more than one slots. Because TowerBFT is not like Alpenglow, in some cases the cluster might have multiple forks competing, so it might be your neighbor's current voting slot is either before or after your own.

Have you tried this on the invalidator?

Reply:
We have duplicate block attack on invalidator so you can see different forks. I'm just wondering if the cluster doesn't confirm at the same speed how much useful data would we get. (e.g. someone in Singapore is always behind everyone else by 2 slots, so we have no data for it)

Reply:
Well this test is set up to tolerate a fair bit of offset between validators. Also the tower votes are only initiating AG consensus for a given slot number, after that it is 100% independent. So this should be able to survive quite a bit of timing offset while providing somewhat useful data. At least that is the idea.